lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+
+    command = pkg.name
+    header = pkg.name
+    library = pkg.name
+
+    # Avoid double 'lib' for packages whose names already start with lib
+    if not pkg.name.startswith('lib'):
+        library = 'lib' + library
+
+    dso_suffix = '.dylib' if sys.platform == 'darwin' else '.so'
+    chmod = which('chmod')
+
+    # Install fake command
+    mkdirp(pkg.prefix.bin)
+    touch(os.path.join(pkg.prefix.bin, command))
+    chmod('+x', os.path.join(pkg.prefix.bin, command))
+
+    # Install fake header file
+    mkdirp(pkg.prefix.include)
+    touch(os.path.join(pkg.prefix.include, header + '.h'))
+
+    # Install fake shared and static libraries
+    mkdirp(pkg.prefix.lib)
+    for suffix in [dso_suffix, '.a']:
+        touch(os.path.join(pkg.prefix.lib, library + suffix))
+
+    # Install fake man page
+    mkdirp(pkg.prefix.man.man1)
+
+    packages_dir = spack.store.layout.build_packages_path(pkg.spec)
+    dump_packages(pkg.spec, packages_dir)
+
+
+def _get_bootstrap_compilers(pkg):
+    """
+    Retrieve uninstalled compilers
+
+    Checks Spack's compiler configuration for a compiler that
+    matches the package spec. If none are configured, installs and
+    adds to the compiler configuration the compiler matching the
+    CompilerSpec object.
+
+    Args:
+        pkg (Package): the package that may need its compiler installed
+
+    Return:
+        (list) list of tuples, (spack.package.PackageBase, bool), for
+            concretized compiler-related packages that need to be installed
+            and bool values specify whether the package is the bootstrap
+            compiler
+            (``True``) or one of its dependencies (``False``)
+    """
+    tty.debug('Bootstrapping {0} compiler for {1}'
+              .format(pkg.spec.compiler, pkg.unique_id))
+    compilers = spack.compilers.compilers_for_spec(
+        pkg.spec.compiler, arch_spec=pkg.spec.architecture)
+    if not compilers:
+        dep = spack.compilers.pkg_spec_for_compiler(pkg.spec.compiler)
+        dep.architecture = pkg.spec.architecture
+        # concrete CompilerSpec has less info than concrete Spec
+        # concretize as Spec to add that information
+        dep.concretize()
+        packages = [(s.package, False) for
+                    s in dep.traverse(order='post', root=False)]
+        packages.append((dep.package, True))
+        return packages
+
+    return []
+
+
+def _hms(seconds):
+    """
+    Convert seconds to hours, minutes, seconds
+
+    Args:
+        seconds (int): time to be converted in seconds
+
+    Return:
+        (str) String representation of the time as #h #m #.##fs
+    """
+    m, s = divmod(seconds, 60)
+    h, m = divmod(m, 60)
+
+    parts = []
+    if h:
+        parts.append("%dh" % h)
+    if m:
+        parts.append("%dm" % m)
+    if s:
+        parts.append("%.2fs" % s)
+    return ' '.join(parts)
+
+
+def _installed_from_cache(pkg, cache_only, explicit):
+    """
+    Install the package from binary cache
```

Comment:
I struggled a bit with naming that and the `_try_install_from_binary_cache` function since both return a `bool` to indicate success/failure.

Changed to `_install_from_cache`.